### PR TITLE
feat: customize keyboard shortcuts

### PR DESCRIPTION
**Pull request by MayurSMahajan**
This feature will let users customize keyboard shortcuts to their preferred key combinations. 

Solves:
#1882 
#1739 
#1985

### CODE REVIEW

**Comment by LucasXu0** [comment]
Hey, @MayurSMahajan. There're some analytical issues. Please take a look. https://github.com/AppFlowy-IO/AppFlowy/actions/runs/4526011004/jobs/7970889839?pr=2112

File: frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:appflowy/startup/tasks/prelude.dart';
+import 'package:appflowy/workspace/application/settings/shortcuts/shortcuts_model.dart';
+import 'package:appflowy_editor/appflowy_editor.dart';
+
+class SettingsShortcutService {
+  late File file;
+  bool shouldWaitForInitializing = true;
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:10]:
```suggestion
  final _initCompleter = Completer<void>();
```

File: frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+  SettingsShortcutService({File? passedFile}) {
+    if (passedFile == null) {
+      _initializeService();
+    } else {
+      file = passedFile;
+      shouldWaitForInitializing = false;
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:17]:
```suggestion
      _initCompleter.complete();
```

File: frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+  }
+
+  Future<void> _initializeService() async {
+    Directory flowyDir = await appFlowyDocumentDirectory();
+    file = await File('${flowyDir.path}/shortcuts/shorcuts.json')
+        .create(recursive: true);
+  }
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:25]:
```suggestion
  Future<void> _initializeService() async {
    Directory flowyDir = await appFlowyDocumentDirectory();
    file = File('${flowyDir.path}/shortcuts/shorcuts.json')
        .createSync(recursive: true);
    _initComplter.complete();
  }
```

File: frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  Future<void> saveShortcuts(List<ShortcutEvent> currentShortcuts) async {
+    file = await file.writeAsString(jsonEncode(currentShortcuts.toJson()),
+        flush: true);
+  }
+
+  Future<List<ShortcutEvent>> loadShortcuts() async {
+    if (shouldWaitForInitializing) {
+      await _initializeService();
+      shouldWaitForInitializing = false;
+    }
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:36]:
```suggestion
    await _initCompleter.future;

```

File: frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_customize_shortcuts_view.dart [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+            ),
+          ],
+        ),
+        const SizedBox(height: 8),
+        Expanded(
+          child: ListView.builder(
+            itemCount: shortcuts.length,
+            itemBuilder: ((context, i) {
+              return ShortcutsListTile(shortcutEvent: shortcuts[i]);
+            }),
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_customize_shortcuts_view.dart:76]:
```suggestion
            itemBuilder: (context, index) => ShortcutsListTile(shortcutEvent: shortcuts[index]),
```

File: frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_customize_shortcuts_view.dart [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+      command += 'shift+';
+    }
+    if (isMetaPressed) {
+      command += 'meta+';
+    }
+    String keyPressed = keyToCodeMapping.keys.firstWhere(
+        (k) => keyToCodeMapping[k] == logicalKey.keyId,
+        orElse: () => '');
+    command += keyPressed;
+    return command;
```
Comment (Xazin) [inline frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_customize_shortcuts_view.dart:177]:
```suggestion
    final keyPressed = keyToCodeMapping.keys.firstWhere(
      (k) => keyToCodeMapping[k] == logicalKey.keyId,
      orElse: () => '',
    );
    
    return command += keyPressed;
```

File: frontend/appflowy_flutter/test/bloc_test/shortcuts_test/shortcuts_cubit_test.dart [hunk 7]
```diff
@@ -0,0 +1,115 @@
+// ignore_for_file: prefer_const_constructors
```
Comment (Xazin) [inline frontend/appflowy_flutter/test/bloc_test/shortcuts_test/shortcuts_cubit_test.dart:1]: Why this ignore?

File: frontend/appflowy_flutter/test/unit_test/settings/shortcuts/settings_shortcut_service_test.dart [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+{
+"key":"End",
+"command":"alt+end"
+},
+{
+"key":"Delete Text",
+"command":"alt+delete"
+}
+]
+}""";
```
Comment (Xazin) [inline frontend/appflowy_flutter/test/unit_test/settings/shortcuts/settings_shortcut_service_test.dart:36]: Could be prettier for readability

**Review by Xazin — COMMENTED** [review]
Some minor comments from my perspective.

File: frontend/appflowy_flutter/test/unit_test/settings/shortcuts/settings_shortcut_service_test.dart (hunk 8 above)
Comment (MayurSMahajan) [reply frontend/appflowy_flutter/test/unit_test/settings/shortcuts/settings_shortcut_service_test.dart:36]: Hey @Xazin, thank you, I wanted to make it more readable as well. But if I insert spaces and linebreaks, then they get recorded in the string and thus corrupt the json. So basically then the jsonDecode method doesn't work. Do you have any other idea to make it prettier for readability

File: frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart (hunk 1 above)
Comment (MayurSMahajan) [reply frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:10]: got to learn something new today, thanks @Xazin!

File: frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart (hunk 3 above)
Comment (Xazin) [reply frontend/appflowy_flutter/lib/workspace/application/settings/shortcuts/settings_shortcuts_service.dart:25]: Updated the suggestion! :+1:

**Comment by MayurSMahajan** [comment]
There is something I want your opinion guys, 
Right now, I am not checking for duplicate shortcuts, 

so for instance: 
if someone assigns the 'ctrl+c' to **cut** operation, I don't check if it is already used by something else (in this case it is used by **copy** operation). It just assigns it without any checking. 

So my question is, how do we want to tackle this, we could:
1. prevent the user from assigning an existing shortcut.
2. swap the shortcuts. ( I think this could be confusing for the user.)
3. prompt the user that the shortcut is already being used and if they still want to reassign it.

cc: 
@annieappflowy 
@LucasXu0 
@Xazin

**Comment by Xazin** [comment]
> There is something I want your opinion guys, Right now, I am not checking for duplicate shortcuts,
> 
> so for instance: if someone assigns the `ctrl+c` to **cut** operation, I don't check if it is already used by something else (in this case it is used by **copy** operation). It just assigns it without any checking.
> 
> So my question is, how do we want to tackle this, we could:
> 
> 1. prevent the user from assigning an existing shortcut.
> 2. swap the shortcuts. ( I think this could be confusing for the user.)
> 3. prompt the user that the shortcut is already being used and if they still want to reassign it.
> 
> cc: @annieappflowy @LucasXu0 @Xazin

Number 3 is the correct choice in my opinion, sometimes you want to use different keybinds for a command, as you reassign it the message saying "The keybinding CTRL + C is already used to perform COPY, reassigning CTRL + C to CUT will remove this keybinding from COPY. [Reassign] [Cancel]" is actually really useful when it mentions what command you lose. Then you can quickly find that command and reassign another keybind there.

**Comment by LucasXu0** [comment]
@MayurSMahajan Would you mind writing a detailed technical document about your design? We want to make discuss your design before merging it.

**Comment by MayurSMahajan** [comment]
> @MayurSMahajan Would you mind writing a detailed technical document about your design? We want to make discuss your design before merging it.

Sure

**Pull request closed without merging**